[Feature](avro) Support Apache Avro file format

DongLiang-0 opened the pull request:
# Proposed changes

#18431

## Problem summary
support read avro file by hdfs() or s3() .

current avro reader only support common data type, the complex data types will be supported later

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

DongLiang-0 committed:
[Feature](avro) Support Apache Avro file format

DongLiang-0 committed:
[Feature](avro) Support Apache Avro file format

sollhui committed:
[Fix](lazy_open) Fix dead lock in lazy open (#19652)

zddr committed:
[improvement](FQDN)Remove unused code (#19638)

zddr committed:
[doc](fqdn)fqdn doc en (#19634)

dutyu committed:
[Fix](multi-catalog) Fix query hms tbl with compressed data files. (#19557)

If a hms table's file format is csv, uncompressed data files may be coexists with compressed data files, so we need to set compressType separately.

reswqa committed:
[feature](tvf) Support compress file for tvf hdfs() and s3() (#19530)

We can support this by add a new properties for tvf, like :

`select * from hdfs("uri" = "xxx", ..., "compress_type" = "lz4", ...)`

User can:

Specify compression explicitly by setting `"compression" = "xxx"`.
Doris can infer the compression type by the suffix of file name(e.g. `file1.gz`)
Currently, we only support reading compress file in `csv` format, and on BE side, we already support.
All need to do is to analyze the `"compress_type"` on FE side and pass it to BE.

stalary committed:
[fix](row-policy) row policy supports external catalog (#19570)

Row policy support external catalog

DongLiang-0 committed:
fix run failed

DongLiang-0 committed:
fix run failed

DongLiang-0 commented:
run buildall

adonis0147 committed:
[chore](third-party) Don't link keyutils to krb5 explicitly (#19632)

We may link system-wide keyutils to krb5 when building krb5 which may introduce an extra dependency to the codebase.

BiteTheDDDDt committed:
[Chore](asan) change asan_suppr from interceptor_via_lib to interceptor_via_fun (#19636)

change asan_suppr from interceptor_via_lib to interceptor_via_fun

DongLiang-0 committed:
fix

DongLiang-0 committed:
fix

DongLiang-0 commented:
run buildall

dataroaring committed:
[fix](delete) notify all when there is no high priority task (#19577)

In somecases high priority threads are waked but normal are not. We
notify_all as a workaround.

wsjz committed:
[feature-wip](multi-catalog)(step1)support connect to max compute (#19606)

Issue Number: #19679

support connect to max compute metadata by odps sdk

dataroaring committed:
[improvement](load) skip compression when memtable is small (#19300)

* [improvement](load) skip compression when memtable is small

* format

hello-stephen commented:
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.42 seconds
 stream load tsv:          433 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230516051549_clickbench_pr_144291.html

eldenmoon committed:
[WIP](row store) two phase opt read row store (#18654)

wzymumon committed:
[vectorized](function) support array_count function (#18557)

support array_count function.
array_count：Returns the number of non-zero and non-null elements in the given array.

zhannngchen committed:
[enhancement](merge-on-write) avoid unecessary pk index iteration (#19620)

xinyiZzz committed:
[Improvement](memory) new page use Allocator (#19472)

xy720 committed:
[docs](struct-type) Add docs for struct type (#19694)

adonis0147 committed:
[chore](third-party) Support specifying packages to build (#19688)

Usage: ./build-thirdparty.sh [options...] [packages...]
  Optional options:
     -j <num>               build thirdparty parallel
     --clean                clean the extracted data
     --continue <package>   continue to build the remaining packages (starts from the specified package)

Examples:
1. Specify packages to build.
    Build gflags, gtest and glog by executing ./build-thirdparty.sh gflags gtest glog.
2. Continue to build the remaining packages.
    Build the remaining packages (starts from sse2neon) by executing ./build-thirdparty.sh --continue sse2neon.

yiguolei committed:
[bugfix](be core) master info is deconstructed before fragment mgr and be will core (#19687)

Co-authored-by: yiguolei <yiguolei@gmail.com>

zy-kkk committed:
[improvement](jdbc catalog) Adapt to hana's special view & Optimize jdbc name format (#19696)

BiteTheDDDDt committed:
[Bug](pipeline) fix distributionColumnIds not updated correct when outputColumnUnique… (#19704)

fix distributionColumnIds not updated correct when outputColumnUnique

xy720 committed:
[docs](struct-type) add docs for struct and named_struct function (#19700)

Yukang-Lian committed:
[doc](retention) optimize retention doc  (#19692)

Yukang-Lian committed:
[doc](doris future) Add mentor doc for doris future in community page (#19690)

fuchanghai committed:
[fix](doc)update readme docs link 404 (#19719)

zddr committed:
[fix](FQDN)fix Checkpoint error (#19678)

Must use Env.getServingEnv() instead of getCurrentEnv(),because here we need to obtain selfNode through the official service catalog.

jacktengg committed:
[fix](ubsan) fix ubsan errors (#19658)

ixu ubsan errors:

doris/be/src/util/string_parser.hpp:275:58: runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'

doris/be/src/vec/functions/functions_comparison.h:214:51: runtime error: addition of unsigned offset to 0x7fea6c6b7010 overflowed to 0x7fea6c6b700c

doris/be/src/vec/functions/multiply.cpp:67:50: runtime error: signed integer overflow: 1295699415680000000 * 0x0000000000015401d0a4cd4890a77700 cannot be represented in type '__int128

doris/be/src/vec/aggregate_functions/aggregate_function_percentile_approx.h:445:73: runtime error: addition of unsigned offset to 0x7feca3343d10 overflowed to 0x7feca3343d08 

doris/be/src/exec/schema_scanner/schema_tables_scanner.cpp:330:24: run

Gabriel39 committed:
[UT](decimalv3) fix FE UT when enable decimal conversion (#19701)

Gabriel39 committed:
[minor](hash table) parameterize hash table (#19653)

zxealous committed:
[improvement](file cache)Support set min file segment size while use block file cache (#19536)

BiteTheDDDDt committed:
[Bug](planner) fix unassigned conjunct assigned on wrong node (#19672)

* fix unassigned conjunct assigned on wrong node

morrySnow committed:
[fix](Nereids) decimal compare float should use double as common type (#19710)

bigben0204 committed:
[feat](profile) Add a new rest api to query instance host and ip information for query profile action in branch master(#18668) (#19643)

eldenmoon committed:
[Bug](topn) fix rowid fetcher merge with empty block (#19712)

jackwener committed:
[minor](Nereids): unify name about Transpose. (#19662)

Gabriel39 committed:
[Improvement](string function) optimize substring and in string set (#19257)

* [Improvement](string function) optimize substring and in string set

* update

nextdreamblue committed:
[fix](functions) fix least/greatest function coredump bug (#19462)

 fix least/greatest function coredump bug

yongjinhou committed:
http interfaces between FEs are not redirected (#19590)

morrySnow committed:
[fix](Nereids) not fallback correctly when do forward (#19675)

starocean999 committed:
[fix](planner)just return non-empty side of ExprSubstitutionMap if one of ExprSubstitutionMap is empty (#19600)

LiBinfeng-01 committed:
[Fix](Nereids) Fix minidump connect context loading and concurrency bug (#19578)

There are two problems of mini dump:
1、minidump do not load connect context to ThreadInfo, so it can not be get easily
2、minidump write maps with not concurrent protection, so the map size with change when we iterating map iterator

Solution:
1、loading connect context to minidump threading
2、use immutable map copy a new map before we actually doing iteration

BiteTheDDDDt committed:
[Chore](function) clean some unused function symbols (#19649)

clean some unused function symbols

dataroaring committed:
[fix](load) donot hang publish due to compaction signal (#19440)

dataroaring committed:
[chore](config) ignore_eovercrowded to be true by default (#19282)

morrySnow committed:
[fix](Nereids) result error when do agg with distinct under pipeline (#19735)

liaoxin01 committed:
[fix](merge-on-write) fix that delete bitmap calculation error when clone tablet (#19713)

xiaokang committed:
[bugfix](inverted index) Fix mv inheriting unexpectedly inverted index of base table (#19722)

Commit:
[improvement](tracing) Remove useless span levels from be side tracing (#19665)

1. Remove an exec node method corresponding to a span and replace it with an exec node corresponding to a span;
2. Fix some problems with tracing in pipeline.

morrySnow committed:
[refactor](Nereids) refactor adjust nullable rule as a custom rewriter (#19702)

use custom rewriter to do adjust nullable to avoid nullable changed in expression but not changed in output

morrySnow committed:
[test](Nereids) diable Nereids explicitly on explain case in regression test (#19744)

Reminiscent committed:
[Fix](Nereids) check the tableName in catalog (#19695)

# Proposed changes
In the nereids. Before this PR: when we access some unexists tables. It will report the exception as follows:
```
mysql> select * from tt;
ERROR 1105 (HY000): errCode = 2, detailMessage = Unexpected exception: null
```

After this PR, it will get the following results:
```
mysql> select * from tt;
ERROR 1105 (HY000): errCode = 2, detailMessage = Unexpected exception: Table [tt] does not exist in database [default_cluster:test].
```

## Problem summary
It is because in this [function](https://github.com/Reminiscent/doris/blob/f5af07f7b218ad2c1266626970b31dd9a9ee72d4/fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java#L328), we ignore the exception. So the size of `tables` in `CascadesContext` is zero not null. So we can only get null after `table = cascadesContext.getTableByName(tableName);`.

sohardforaname committed:
[Fix](planner) fix literal type incompatible after fold constant by be. (#19190)

LiBinfeng-01 committed:
[Fix](Nereids) fix minidump parameter name and double not a number serialize bug (#19635)

Change nereids minidump switch from "Dump_nereids" to "enable_minidump" which is more exactly and neat. Also fix bug of Double.NaN (not a number) serialize bug when doing column statistic serialization.

morningman committed:
[fix](meta) fix upgrade failed on FE meta from 1.2 (#19674)

Introduced from #19355.
We need to keep OP_CREATE_CLUSTER so that Doris can be upgraded from 1.2.x.
This OP type should be removed after 3.0

AshinGau committed:
[fix](multi-catalog) verify the precision of datetime types for each data source (#19544)

Fix threes bugs of timestampv2 precision:
1. Hive catalog doesn't set the precision of timestampv2, and can't get the precision from hive metastore, so set the largest precision for timestampv2;
2. Jdbc catalog use datetimev1 to parse timestamp, and convert to timestampv2, so the precision is lost.
3. TVF doesn't use the precision from meta data of file format.

BePPPower committed:
[Bugfix](Jdbc Catalog) fix data type mapping of SQLServer Catalog (#19525)

We map `money/smallmoney` types of SQLSERVER into decimal type of doris.

sohardforaname committed:
[Feature](Nereids) add executable function to support fold constant for functions (#18209)

1. Add date-time functions for fold constant for Nereids.
This is the list of executable date-time function nereids supports up to now:
- now()
- now(int)
- current_timestamp()
- current_timestamp(int)
- localtime()
- localtimestamp()
- curdate()
- current_date()
- curtime()
- current_time()
- date_{add/sub}(),{years/months/days/hours/minutes/seconds}_{add/sub}()
- datediff()
- {date/datev2}()
- {year/quarter/month/day/hour/minute/second}()
- dayof{year/month/week}()
- date_format()
- date_trunc()
- from_days()
- last_day()
- to_monday()
- from_unixtime()
- unix_timestamp()
- utc_timestamp()
- to_date()
- to_days()
- str_to_date()
- makedate()

2. solved problem:
- enable datev2/datetimev2 default.
- refactor Nereids foldConstantOnFE and support fold nested expression.
- separate the executable into multi-files for easily-reading and adding new functions

Commit:
[Improve](complex-type)add complex type support unique table with regress test #19751

add complex type support unique table with regress test
struct / map / array now support unique table but no regress test

wsjz committed:
[fix](multi-catalog)fix iceberg catalog display type #19728

zddr committed:
[fix](checkpoint)fix Checkpoint error when http server is not ready #19699

HappenLee committed:
[Log](expr) add DCHECK info for expr close DCHECK (#19683)

Jibing-Li committed:
[Fix](multi catalog, nereids)Fix FileQueryScanNode couldn't filter partition in nereids planner bug (#19564)

Nereids planner add conjuncts to ScanNode after call finalize, this may cause external table scan node fail to filter 
useless partition, because external table do the partition prune in the finalize method.

This pr is to fix this bug. In the rewrite stage, pass the conjuncts to LogicalFileScan object, and eventually pass to 
ScanNode while creating it. So that the ScanNode could use the conjuncts while doing finalize. 

Why not doing the partition prune in the LogicalFileScan like LogicalOlapScan doing? 

Because Iceberg api doesn't have the partition concept, it just accept a list of Conjuncts, 
so it's easier to pass the conjuncts to ScanNode (Hive, Icegerg, Hudi...) and doing the partition prune in there.

morningman committed:
[doc](catalog) optimize catalog doc (#19601)

dataroaring committed:
[improvement](load) do not create pthread in tablet_sink (#19465)

add bvar stat for streamload.

herry2038 committed:
[feature](compact) Duplicate with no keys tables compaction coredump (#19490)



Co-authored-by: yuxianbing <yuxianbing@yy.com>

catpineapple committed:
[doc](fix) cold hot separation doc fix (#19757)

zy-kkk committed:
[improvement](storage) let the `storage_page_cache_shard_size` conf be rounded up to a power of two (#19639)

Commit:
[fix](tvf) fix the inconsistency between tvf backends function and show backends result (#19697)

zy-kkk committed:
[fix](catalog) fix the `include_database_list` not in effect (#19589)

nextdreamblue committed:
[enhancement](schema) dynamic_partition.time_unit support year (#19551)

dynamic_partition.time_unit support year

lvshaokang committed:
[Enhancement](broker-load) broker load show stmt support display cluster name if specified (#19392)

freemandealer committed:
[fix](sink) disable lazy-open partition by default (#19769)

Signed-off-by: freemandealer <freeman.zhang1992@gmail.com>

HappenLee committed:
[Bug](pipeline) RegressionTest failed release resouce cause DCHECK failed #19773

caoliang-web committed:
[typo](doc)Example of adding stream load import to current time (#19760)

jackwener committed:
[log](Nereids): add more debug info when check logicalproperties. (#19763)

figurant committed:
[typo](docs) spark load example revise and some other revise (#19756)

Co-authored-by: zhuwei <zhuwei8421@gmail.com>

gnehil committed:
[typo](doc)fix audit plugin and spark load kerberos for yarn #19749

CalvinKirs committed:
[typo](docs)Fix typo in monitor-alert (#19743)

gwxog committed:
[typo](doc)fix typo in BROKER-LOAD.md #19737

HappenLee committed:
[Bug](pipeline) RegressionTest failed release resouce cause DCHECK failed #19772

liaoxin01 committed:
[enhancement](merge-on-write) Avoiding unnecessary primary key index traversal (#19746)

airborne12 committed:
[Chore](inverted index) change Status::EndOfFile to just logging info, remove useless print (#19721)

xinyiZzz committed:
[fix](memory)  remove auto option in memory config and optimize memtracker logs #19706

fix mem_limit default value
memory_gc_sleep_time_s to memory_gc_sleep_time_ms
LoadChannelMgr::_handle_mem_exceed_limit process_mem_limit to process soft mem limit
fix query mem tracker print

lsy3993 committed:
[typo](docs)add new describe for grant (#19656)

Commit:
[Doc](map-type) add map sql doc  (#19612)

yuanyuan8983 committed:
[typo](docs) modify the dynamic table document (#19750)

xiaokang committed:
[feature](like) make like regexp used as a sql function (#19755)

HappenLee committed:
[pipeline](CTE) Support multi stream data sink in pipeline (#19519)

jackwener committed:
[git](Nereids): ignore apache parquet. (#19765)

chenlinzhong committed:
[improvement](load)disable  shrink memory by default (#19714)

disable shrink memory by default, it becomes very slow when importing large amounts of data
you can turn on If you think it's necessary

starocean999 committed:
[fix](nereids) fix some nereids bugs (#19711)

1. add json_unquote and json_extract functions
2. remove mv releated code in visitPhysicalOlapScan
3. forbid bitmap and hll type for topn node's sort exprs
4. HashDistributionInfo of olap scan node should use the slots from output not the full schema
5. SelectMaterializedIndexWithoutAggregate should use the filter node's output together with the predicate to get the correct mv
6. forbid SimplifyArithmeticRule for decimal type
7. make DecimalLiteral's type and value consistent with each other if the value is decimalv2
8. json_array need support empty argument

bobhan1 committed:
[refactor-WIP](TaskWorkerPool) add CreateTableTaskPool class for CREATE_TABLE task (#19734)

yixiutt committed:
[enhancement](compaction) not core when init failed (#19754)

fuchanghai committed:
[improvement-#19739]statement error in data-model (#19740)

Co-authored-by: fuchanghai <‘2875334588@qq.com’>

ZI-MA committed:
[regressiontest](unionall) Regression_test_similar_query_boolean (#19553)

* regression_test_similar_query

* add the ORDER BY

* update ORDER BY to comfirm correctness

---------

Co-authored-by: ZI-MA <chime316@qq.com>

Hastyshell committed:
[fix](load) fix unified load converted failed when forwarding to master (#19779)

yixiutt committed:
[improvement](MOW) use seperated cache for mow pk cache (#19686)

In mow, primary key cache have a big impact on load performance, so we add a new cache type to seperate
it from page cache to make it more flexible in some cases

morningman committed:
[fix](mtmv) fix bug that should not write edit log when replaying alter mv (#19781)

Gabriel39 committed:
[minor](datev2) remove datev2 because datev2 is used by default (#19777)

airborne12 committed:
[Fix](single replica load) add inverted index copy for single replica load (#19663)

* [Fix](single replica load) add inverted index copy for single replica load

Reminiscent committed:
[minimal](Nereids) add more comments for the rewriter (#19788)

Only add some comments to the rewriter. Because it is fewer comments before and it's hard to understand for the newbie.

xinyiZzz committed:
[fix](memory) Print all query/load memory before memory GC when `memory_debug=true` (#19720)

xinyiZzz committed:
[Improvement](memory) faststring use Allocator #19762

After the outer catch exception, faststring resize reserve build may throw a memory alloc failure exception from the Allocator.

Currently page body compress will catch memory alloc failure exception

BiteTheDDDDt committed:
[Chore](build) fix some undefined behavior about incomplete type vector #19753

xiaokang committed:
[feature](jsonb) rename JSONB type name and function name to JSON (#19774)

To be more compatible with MySQL, rename JSONB type name and function name to JSON.

The old JSONB type name and jsonb_xx function can still be used for backward compatibility.

There is a function jsonb_extract remained since json_extract is used by json string function and more work need to change it. It will be changed further.

eldenmoon committed:
[Optimize](row store) optimize serialization and deserialization (#19691)

1. Get DataTypeSerde in advance to avoid get temporary DataTypeSerde iterate each column
2. Iterate the original row once is enoungh for deserializing by introducing a map for record the index of each column's unique id

ZhangYu0123 committed:
[enhence](memory) gc inverted index cache when there is not enough memory (#19622)

Support to gc inverted index cache when there is not enough memory.
previous problem： The inverted index cache (InvertedIndexSearcherCache and InvertedIndexQueryCache) may use 20% memory which can't be released.

morrySnow committed:
[fix](Nereids) fallback not work when cannot parse after forward (#19790)

bobhan1 committed:
[refactor-WIP](TaskWorkerPool) add DropTableTaskPool for DROP_TABLE task (#19793)

dutyu committed:
[Fix](multi-catalog) Fix sync hms event failed. (#19555)

A similar situation with #19344 , because sometimes hms meta info is newer than hms events, if we try to invoke org.apache.doris.datasource.hive.PooledHiveMetaStoreClient#getTable and this table is not exists, some error will throws and this event can not be handled.

Commit:
[improvement](opentelemetry) upgrade opentelemetry jar to v1.26.0 and opentelemetry-cpp to v1.8.3 (#19733)


why upgrade? anything wrong?

Try to fix the problem about opentelemetry::v1::ext::http::client::curl::HttpOperation::Send(), I have updated the pr info.

HappenLee committed:
[Bug](pipeline) RegressionTest failed release resouce cause DCHECK failed (#19783)

RegressionTest failed release resouce cause DCHECK failed

Cai-Yao committed:
[Refactor](spark load) remove parquet scanner (#19251)

yongjinhou committed:
fix select resource groups bug (#19808)

zy-kkk committed:
[fix](jdbc catalog)fixed oceanbase catalog row limit bug (#19796)

zy-kkk committed:
[improvement](jdbc catalog) Add adaptation to Oracle special character  `/` table names (#19809)

adonis0147 committed:
[bug](MTMV) Fix the wrong interpretation for NEVER REFRESH (#19800)

dataroaring committed:
[improvement](stacktrace) do not captute stack trace for txn error codes (#19817)

airborne12 committed:
[Update](build) change clucene from thirdparty to git module (#19352)

yagagagaga committed:
[typo](docs) Add a previously missed configuration item and other note (#19814)

fix enable_query_memory_overcommit config doc

morningman committed:
[fix](regression-test) fix insert overwrite case same db name issue (#19839)

kaka11chen committed:
[Fix](parquet-reader) Fix partition field conjuncts not work. (#19837)

Fix partition field conjuncts not work.
Add predicate_partition_columns in _slot_id_to_filter_conjuncts(single slot conjuncts) to _filter_conjuncts, others should had been added from not_single_slot_filter_conjuncts.

morningman committed:
[deps](libhdfs) update hadoop libhdfs to 3.3.4.1 for doris (#19832)

Commit:
[improvement](profile) add profile counter 'BytesSent' for VDataBufferSender (#19826)

morningman committed:
[chore](build) avoid generating generated code every time (#19813)

When calling generated-source.sh in build.sh, not to remove the gensrc/build dir.

xinyiZzz committed:
[fix](profile) Fix pipeline load channel profile #19828

englefly committed:
[improve](nereids) check be status when column stats is unknown #19742

when forbid_unknown_col_stats is open and some column stats is unknown,
we will check the be status by StatisticsUtil.statsTblAvailable(), and report error according to be status.

morningman committed:
[minor](log) add details for unqueryable replicas (#19792)

Add a new FE config: show_details_for_unaccessible_tablet.
Default is false, when set to true, if a query is unable to select a healthy replica,
the detailed information of all the replicas of the tablet including the specific reason why they are unqueryable,
will be printed out.

englefly committed:
[fix](nereids) avoid 0 row count in stats derive #19640

row count of join estimation is at least 1 to make less error propagation.

Kikyou1997 committed:
[fix](test) Comment unstable stats test #19729

Commit:
[fix](tracing) fix the issue that a trace may track multiple queries (#19804)

englefly committed:
[feature](nereids) set proper min/max value for column stats when minExpr/maxExpr is not avialable #19673

zclllyybb committed:
[chore](build) Support specifying output path #19669

now we could

./build.sh --be --fe --output PATH_TO_BINARY_YOU_LIKE
did not modify the default value

yixiutt committed:
[Feature](planner) use partial update in update from & delete from (#19262)

airborne12 committed:
[Chore](build) fix typo in build.sh (#19846)

eldenmoon committed:
[explain](point query) modify explain for SHORT-CIRCUIT query (#19820)

yiguolei committed:
[improvement](session variable) add max execution time session variabe like mysql and add setter attributes in variables (#19759)

1. add session variable max_execution_time to an alias of query timeout, if user set max_execution_time, the query timeout will be modified too.
2. add a setter attribute to session variable, so that we could add some logic in setter method instead of field reflection.

morrySnow committed:
[fix](Nereids) normalize repeat generate push down project with error nullable (#19831)

jackwener committed:
[minor](Nereids): add toString() for LogicalProperties (#19851)

Gabriel39 committed:
[Bug](DecimalV3) fix decimalv3 functions (#19801)

amorynan committed:
[FIX](Array)fix be master comapitible with fe1.2 #19850

when upgrade be first , fe is 1.2 , array contains_nulls is set not in thrift ,so would has core in be
Issue Number: close #xxx

DongLiang-0 committed:
[typo](doc)flink and spark connector remove thrift compiled documents (#19794)

* [typo](doc)flink and spark connector remove thrift compiled documents

* delete enable_http_server_v2

liangjiawei1110 committed:
[Enhancement](meta) Show remote data usage via SHOW DATA #19533 (#19752)

* [Enhancement](meta) Show remote data usage via SHOW DATA #19533

* [fix] correct some unit test results

airborne12 committed:
[Fix](inverted index) add datev2/datetimev2 for inverted index column type (#19845)

When we try to query array of datetimev2 column by inverted index, it returns an error like this:

CREATE TABLE `nested` (
 `qid` bigint(20) NULL,
 `tag` array<text> NULL,
 `creationDate` datetime NULL,
 `title` text NULL,
 `user` text NULL,
 `answers.user` array<text> NULL,
 `answers.date` array<datetimev2(0)> NULL,
 INDEX tag_idx (`tag`) USING INVERTED PROPERTIES("parser" = "english") COMMENT '',
 INDEX creation_date_idx (`creationDate`) USING INVERTED COMMENT '',
 INDEX title_idx (`title`) USING INVERTED COMMENT '',
 INDEX user_idx (`user`) USING INVERTED COMMENT '',
 INDEX answers_user_idx (`answers.user`) USING INVERTED COMMENT '',
 INDEX answers_date_idx (`answers.date`) USING INVERTED COMMENT ''
) ENGINE=OLAP
DUPLICATE KEY(`qid`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`qid`) BUCKETS 18
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"storage_format" = "V2",
"compression" = "ZSTD",
"light_schema_change" = "true",
"dynamic_schema" = "true",
"disable_auto_compaction" = "false"
); 

mysql> select * from nested.nested where tag match 'java' and `answers.date` element_le '2012-04-08T21:15:33.873Z' limit 10;
ERROR 1105 (HY000): errCode = 2, detailMessage = no function found for MATCH_ELEMENT_LE,`answers.date` MA

xinyiZzz committed:
[fix](memory) Remind log if `vm/overcommit_memory`=2 when be start (#19795)

Expect vm overcommit memory value to be 1, system will no longer throw bad_alloc, memory alloc are always accepted,
memory limit check is handed over to Doris Allocator, make sure throw exception position is controllable,
otherwise bad_alloc can be thrown anywhere and it will be difficult to achieve exception safety.

zddr committed:
[improvement](conf)Add an example of directly specifying an IP address #19860

morrySnow committed:
[fix](Nereids) drop table before create table in window function test (#19857)

Gabriel39 committed:
[Bug](decimal) fix variance_samp and avg_weighted #19861

zddr committed:
[feature-wip](MTMV) support multi catalog (#19854)

* mtmv support multi catalog

* mtmv support multi catalog

eldenmoon committed:
[Doc](point query) Modify the document to add constraints and limitations for optimizing high-concurrency point queries. (#19852)

amorynan committed:
[Improve](complex-type)improve array/map/struct creating and function with decimalv3 (#19830)

adonis0147 committed:
[chore](build) Fix the compilation errors on macOS (arm64) (#19859)

Some errors raise when building the codebase on macOS (arm64).

jackwener committed:
[enhance](Nereids): Pushdown Filter Through Project in Post Processor. (#19873)

Originally, PushdownFilterThroughProject is in CBO phase, but it will increase Memo size.
So, we move it into PostProcessor

zy-kkk committed:
[improvement](jdbc catalog) set oceanbase mysql mode jdbc param `useCursorFetch` default true (#19856)

LiBinfeng-01 committed:
[Fix](Nereids) fix serialize colocate table index concurrent bug (#19862)

When doing serialization of minidump input, we can find that when serializing colocate table index, the size and entry get by the hash map always unmatched when concurrent occur. So a write lock be added to ensure concurrency.

Gabriel39 committed:
[decimalv3](function) support function width_bucket (#19806)

xinyiZzz committed:
[fix](memory)page no use Allocator, avoid ckbench oom (#19877)

jackwener committed:
[fix](community): remove some collaborators. (#19884)

jackwener committed:
[fix](Nereids): BuildAggForUnion forgot to convert Qualifier Type. (#19883)

HappenLee committed:
[Bug][pipeline] Fix regression tpcds failed in nereid planner (#19885)

zddr committed:
[doc](multi-catalog)change hms event doc (#19886)

change hms event doc
need restart more service

caoliang-web committed:
Modify the usage of Doriswriter parameters (#19875)

yuanyuan8983 committed:
Modify the jdk version requirements in the documentation (#19874)

catpineapple committed:
update create table doc (#19871)

xiaokang committed:
add testcases for inverted index on different datatypes (#19843)

bingquanzhao committed:
[Doc](fix) Fixed two documentation errors (#19879)

xinyiZzz committed:
[fix](ubsan) UBSAN avoid thread local switch

zddr committed:
[improvement](multi-catalog)hms sync event log more info #19887

zddr committed:
[improvement](FQDN) broker support fqdn (#19821)

1.broker support fqdn
2.change 'master_only' attr of 'enable_fqdn_mode'

whutpencil committed:
[Enhancement] show total transactions in show proc "/transactions" (#19492)

In a scenario where multiple DBs are simultaneously imported with high concurrency, a significant number of transactions will be generated. Without a summary field, we cannot clearly see how many transactions there are in the current cluster. Therefore, I have enhanced this point.

```
mysql> show proc "/transactions";
+-------+-----------------------------------+-----------------------+
| DbId     | DbName                                             | RunningTransactionNum |
+-------+-----------------------------------+-----------------------+
| 10002   | default_cluster:xxxx                         | 0                                     |
| 14005   | default_cluster:__internal_schema | 0                                     |
| Total     |  2                                                         | 0                                     |
+-------+-----------------------------------+-----------------------+
3 rows in set (0.02 sec)
```

myfjdthink committed:
[feature](iceberg) Support read iceberg data on gcs (#19815)

wangbo committed:
[test](executor)add crud regression test for resource group (#19659)

dd crud regression test for resource group (#19659)

morningman committed:
[minor](clone) add more debug log for tablet scheduler (#19892)

Sometimes I find that the tablet scheduler can not schedule tablet, and with no more info for debugging.
So I add some debug log for this process.
No logic is changed.

morningman committed:
[refactor](cluster)(step-2) remove cluster related to Backend  (#19842)

Commit:
[feature-wip](resouce-group)  Supports memory soft isolation of resource group (#19802)

create resource groups name properties(
    'enable_memory_overcommit' = 'true' // whether to enable memory soft isolation
)

xinyiZzz committed:
[revert](memory) revert page no use Allocator && default disable ChunkAllocator (#19905)

default chunk allocator reserve is 0. At this time, it is meaningless to enable chunk allocator, it will only waste memory.

ZhangYu0123 committed:
[fix](config) fix memory config enable_query_memroy_overcommit spell problem #19898

morningman committed:
[refactor](hudi) remove hudi external table (#19908)

Hudi external table is deprecated since 1.2.
We should remove it now.
Recommend to use "multi-catalog" feature to connect to Hudi.

User can not create Hudi external table.
When restarting FE, all hudi external table will still be replayed but can not be read. And when doing checkpoint, all these tables will be discarded.

Commit:
[FIX](serde)pb ut is not stable #19907

Hastyshell committed:
[typo](docs) replace the wrong field "json_paths" in s3 tvf related docs with the correct one "jsonpaths" #19900

englefly committed:
[opt](nereids)compute runtime filter size by the ndv of build side (#18803)

fsilent committed:
[improvement](libhdfs) Use keytab and principal to login kerberos (#19841)

User keytab and princpal to login kerberos.
And user does not need to execute kinit manually anymore.

BiteTheDDDDt committed:
[Bug](function) fix function in get wrong result when input const column (#19791)

fix function in get wrong result when input const column

BiteTheDDDDt committed:
[Feature](aggregation) add agg_state define and ddl support (#19824)

add agg_state define and ddl support

zddr committed:
[doc](fqdn)broker fqdn doc #19910

niebayes committed:
[typo](doc)fix typo in readme #19895

miliseconds -> milliseconds

BiteTheDDDDt committed:
[Bug](function) make VcompoundPred optimization work well (#19870)

make VcompoundPred optimization work well
#19818 this pr try to enable VcompoundPred optimization but get wrong result on tpcds q28.
The reason is some nullable logic on mysql need special handling.

mysql [regression_test_tpcds_sf1_p1]>select null and false;
+----------------+
| NULL AND FALSE |
+----------------+
|              0 |
+----------------+
1 row in set (0.00 sec)

mysql [regression_test_tpcds_sf1_p1]>select null and true;
+---------------+
| NULL AND TRUE |
+---------------+
| NULL          |
+---------------+
1 row in set (0.00 sec)

mysql [regression_test_tpcds_sf1_p1]>select null or false;
+---------------+
| NULL OR FALSE |
+---------------+
| NULL          |
+---------------+
1 row in set (0.00 sec)

mysql [regression_test_tpcds_sf1_p1]>select null or true;
+--------------+
| NULL OR TRUE |
+--------------+
|            1 |
+--------------+
1 row in set (0.00 sec)

Commit:
[Improve](struct)improve struct support into outfile (#19894)

support select into outfile for struct type

kaka11chen committed:
[Fix][Refactor] Fix 'not member call on null pointer of type 'doris::TextConverter' error in ubsan env and refactor text converter. (#19849)

Fix 'not member call on null pointer of type doris::TextConverter' error in ubsan env and refactor text converter.

Gabriel39 committed:
[Improvement](function) support decimalv3 for function `least` and `greatest` (#19931)

ByteYue committed:
[enhance](fs) use bvar to monitor s3 file reader& writer (#19607)

remove useless prefix

sollhui committed:
[Fix](lazy_open) Fix lazy open null point (#19829)

AshinGau committed:
[fix](multi-catalog) JDBC Catalog Unknown UNSIGNED type of mysql, type: [DOUBLE] (#19912)

xy720 committed:
[docs](struct-map-type) update user docs for struct and map type (#19939)

We have already supported decimalV3.

jackwener committed:
[fix](Nereids): commute non-inner join for DPHyp (#19929)

Jibing-Li committed:
[Fix](multi catalog)Change all partition names to lower case (#19816)

Iceberg table partition name may contain upper case characters, for example: City=xxx, Nation=xxx.
But in Doris, all column names are in lower case. Here we transfer the partition name to lower case to keep consist with column name.

zclllyybb committed:
[fix](testcase) fix wrong use of same table in different case

zhannngchen committed:
[fix](merge-on-write) return error st if check_pk_in_pre_segments failed (#19736)

yixiutt committed:
[enhancement](page cache) use separate pk index cache (#19864)

sollhui committed:
[test](lazy open) add lazy open test (#19896)

CalvinKirs committed:
[refactor](routineload)Refactored routineload to improve scalability (#19834)

- The data source parameters are sunk into the specific data source class
- Simplify some code logic to reduce code complexity
- Provide a data source factory class to extract public logic
- Code that removes tests from production code. We should not include code for testing purposes in any production code.

morrySnow committed:
[test](Nereids) enable all query in tpcds empty table test (#19945)

morrySnow committed:
[fix](Nereids) join reorder lead to circle in memo (#19935)

If we have join as the root node, then after some join reorder join, the root Group in Memo will have a GroupExpression including LogicalProject as its plan and the children is its ownerGroup.
This PR add a rewrite rule to ensure we have a Project on the top of the top Join of plan to avoid circle in Memo.

wsjz committed:
[fix](s3)fix s3 resource check (#19933)

fix s3 resource check:

ERROR 1105 (HY000): Unexpected exception: org.apache.doris.common.DdlException: errCode = 2, detailMessage = Missing [AWS_ACCESS_KEY] in properties.

we should use new properties to check s3 available

gitccl committed:
[Enhancement](scanner) allocate blocks in scanner_context on demand and free them on close (#19389)

Firstly, to reduce memory usage, we do not pre-allocate blocks, instead we lazily allocate block when upper call get_free_block. And when upper call return_free_block to return free block, we add the block to a queue for memory reuse, and we will free the blocks in the queue when the scanner_context was closed instead of destructed.
Secondly, to limit the memory usage of the scanner, we introduce a variable _free_blocks_capacity to indicate the current number of free blocks available to the scanners. The number of scanners that can be scheduled will be calculated based on this value.

ssb flat test
previous
lineorder 1.2G:
load time: 3s, query time: 0.355s
lineorder 5.8G:
load time: 330s, query time: 0.970s
load time: 349s, query time: 0.949s
load time: 349s, query time: 0.955s
load time: 360s, query time: 0.889s (pipeline enabled)
after
lineorder 1.2G:
load time: 3s, query time: 0.349s
lineorder 5.8G:
load time: 342s, query time: 0.929s
load time: 337s, query time: 0.913s
load time: 345s, query time: 0.946s
load time: 346s, query time: 0.865s (pipeline enabled)

xy720 committed:
[Bug](schema-change) fix varchar can not change to datev2 #19952

CalvinKirs committed:
[Chore](docs)Remove empty pages to avoid ambiguity (#19953)

* [Chore](docs)Remove empty pages to avoid ambiguity

* update slidebar

xy720 committed:
[bug](compile) fix fe compile error #19946

Fix fe maven package has a version conflict for package grpc-core.

zddr committed:
[doc](multi-catalog)Supplementary FAQ (#19911)

* catalog doc

* catalog doc

Tanya-W committed:
[fix](schema change) Change table state to NORMAL by SchemaChangeJob instead of SchemaChangeHandler (#19838)

fix problem:
If there is an unfinished schema change job (job-2), and before this time, another schema change job (job-1) of the same table has been finished.
Then restart fe, will replay edit log (pending log and waiting_txn log) for job-2, and the table's state is set to SHCEMA_CHANGE, but when loadAlterJob after replayJournal, will add job-1 to schema change handler, and then run the job-1 will set the table to NORMAL because of job-1 is done, but at this point, the job-2 is doing runWaitingTxnJob, in this function will check table's state, if not normal will throw exception, not change the job's state, and cannot cancel the job because the table is not under schema change.

Gabriel39 committed:
[Bug](decimal) fix `mod` function (#19925)

Bug:
select id, kdcml * ktint, kdcml / ktint, kdcml % ktint from expr_test order by id;
+------+-------------------+-------------------+-----------------------+
| id | kdcml * ktint | kdcml / ktint | kdcml % ktint |
+------+-------------------+-------------------+-----------------------+
| NULL | NULL | NULL | NULL |
| 1 | 24.395 | 24.395 | -4702111234474983.74 |
| 2 | 68.968 | 17.242 | -4702111234474983.74 |
| 3 | 146.268 | 16.252 | -4702111234474983.74 |
| 4 | 275.772 | 17.235 | -4702111234474983.74 |
| 5 | 487.470 | 19.498 | -4702111234474983.74 |
| 6 | 827.244 | 22.979 | -4702111234474983.74 |
| 7 | 1364.860 | 27.854 | -4702111234474983.74 |
| 8 | 2205.928 | 34.467 | -4702111234474983.74 |
| 9 | 3509.595 | 43.328 | -4702111234474983.74 |
| 10 | 5514.790 | 55.147 | -4702111234474983.74 |
| 11 | 8578.988 | 70.900 | -4702111234474983.74 |
| 12 | 13235.484 | 91.913 | -4702111234474983.74 |
| 13 | 24.395 | 24.395 | -4702111234474983.74 |
| 14 | 68.968 | 17.242 | -4702111234474983.74 |
| 15 | 146.268 | 16.252 | -4702111234474983.74 |
| 16 | 275.772 | 17.235 | -4702111234474983.74 |
| 17 | 487.470 | 19.498 | -4702111234474983.74 |
| 18 | 827.244 | 22.979 | -4702111234474983.74 |
| 19 | 1364.860 | 27.854 | -4702111234474983.74 |
| 20 | 2205.928 | 34.467 | -4702111234474983.74 |
| 21 | 3509.595 | 43.328 | -4702111234474983.74 |
| 22 | 5514.790 | 55.147 | -4702111234474983.74 |
| 23 | 8578.988 | 70.900 | -4702111234474983.74 |
| 24 | 13235.484 | 91.913 | -4702111234474983.74 |

DongLiang-0 committed:
get table schema by jni

DongLiang-0 committed:
get table schema by jni

HappenLee committed:
[testcase](test) Fix query stats test may failed (#19958)

englefly committed:
[regression](nereids) tpcds sf100 plan shape regression cases (#19913)

cnissnzg committed:
[performance](load) use vector instead of skiplist when insert agg keys (#19099)

DongLiang-0 committed:
fix avro init method

DongLiang-0 committed:
fix avro init method

DongLiang-0 commented on be/src/vec/exec/scan/avro_reader.cpp:
```diff
@@ -0,0 +1,97 @@
+        case TFileType::FILE_BROKER:
+        case TFileType::FILE_STREAM:
+        default:
+            Status::InternalError("unsupported file reader type: {}", std::to_string(type));
```
Thanks for your suggestion, I have done this.

DongLiang-0 commented on fe/fe-core/src/main/java/org/apache/doris/tablefunction/S3TableValuedFunction.java:
```diff
@@ -79,12 +80,15 @@ public S3TableValuedFunction(Map<String, String> params) throws AnalysisExceptio
         String usePathStyle = tvfParams.getOrDefault(PropertyConverter.USE_PATH_STYLE, "false");
         locationProperties.put(PropertyConverter.USE_PATH_STYLE, usePathStyle);
+        locationProperties.put(S3Properties.VIRTUAL_BUCKET, virtualBucket);
+        locationProperties.put(S3Properties.VIRTUAL_KEY, getVirtualKey());
```
The original `BUCKET` value is not correct. I am not sure if this is a bug. There are quite a lot of places involved in this variable, so I did not change it and created a new `VIRTUAL_BUCKET`.

DongLiang-0 commented on fe/java-udf/pom.xml:
```diff
@@ -45,6 +45,26 @@ under the License.
+            <artifactId>hadoop-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
```
Thanks for your suggestion, I have done this.

ZI-MA committed:
[fix](status) fix function code_as_string to private (#19960)

* rebase

* update format

---------

Co-authored-by: ZI-MA <chime316@qq.com>

GoGoWen committed:
[typo](doc) fix some typo (#19414)

zhannngchen committed:
[doc](merge-on-write) add some notes for using MoW (#19968)

Commit:
[docs](resource-group) add relevant documents to the resource group (#19941)

zclllyybb committed:
[fix](testcases) Fix some unstable testcases. (#19956)

case of test_string_concat_extremely_long_string will exceed our test limit. Move it to p2 so that it will be tested only in SelectDB test environment.
Because we need to keep consistent with MySQL & avoid overflow. the q67 must keep its behavior like now. When we fully apply nereids & decimalV3 then, it will be fixed automatically.
In the parallel test, although all query stats were cleaned, the cases run in parallel will affect this. So we need to use a unique table for query_stats_test
test_query_sys_tables didn't deal with some unstable situations. fixed it.
temporarily disable unstable case analyze_test case for p0.

Tanya-W committed:
[fix](function) fix result column is nullable type when fast execute (#19889)

xinyiZzz committed:
[fix](memory) query check cancel while waiting for memory in Allocator, and optimize log (#19967)

After the query check process memory exceed limit in Allocator, it will wait up to 5s.
Before, Allocator will not check whether the query is canceled while waiting for memory, this causes the query to not end quickly.

DongLiang-0 committed:
format be

DongLiang-0 committed:
format be

DongLiang-0 committed:
Merge remote-tracking branch 'origin/avro-format' into avro-format

# Conflicts:
#	be/src/service/internal_service.cpp

DongLiang-0 committed:
Revert "format be"

This reverts commit 106aa46f999635cf2d818e61bc656362ea246001.
